### PR TITLE
add readme parsing tests

### LINKED ISSUE #5: README parsing tests
This project is currently lacking tests, including for complex parts of the code like README parsing.

Tests should be implemented to primarily ensure that edge cases for parsing build and deploy commands out of Markdown files are covered.

### PR DESCRIPTION
**Pull request by SaketramDurbha**
This PR addresses issue #5 .

- Add readme parsing tests
- Refactor acutal readme parsing code to be more testable

Will rebase and update once PR #22 is merged.

### CODE REVIEW

File: internal/lifecycle/readme.go [hunk 1]
```diff
@@ -42,7 +42,11 @@ var (
 
 	mdCodeFenceStartRegexp = regexp.MustCompile("^\\w*`{3,}[^`]*$")
 
-	errNoREADMECodeBlocksFound = fmt.Errorf("[lifecycle.parseREADME]: no code blocks immediately preceded by %s found", codeTag)
+	errNoREADMECodeBlocksFound   = fmt.Errorf("[lifecycle.parseREADME]: no code blocks immediately preceded by %s found", codeTag)
```
Comment (kurtisvg) [inline internal/lifecycle/readme.go]: We drop the `[funcName]` in another PR I think - should we do so here as well?

Comment (kurtisvg) [reply internal/lifecycle/readme.go]: if these errors aren't being returned in multiple locations, It's probably not worth declaring in advance.

File: internal/lifecycle/readme.go [hunk 2]
```diff
@@ -109,9 +113,19 @@ func parseREADME(filename, serviceName, gcrURL string) (Lifecycle, error) {
...
+
+	l, err := extractLifecycle(scanner, serviceName, gcrURL)
+	if err != nil {
+		return l, err
+	}
+
+	return l, nil
+}
+
+func extractLifecycle(scanner *bufio.Scanner, serviceName, gcrURL string) (Lifecycle, error) {
```
Comment (kurtisvg) [inline internal/lifecycle/readme.go:125]: Should you add a comment to this function?

File: internal/lifecycle/readme_test.go [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+)
+
+func setEnv(e map[string]string) error {
+	for k, v := range e {
+		err := os.Setenv(k, v)
+		if err != nil {
+			return err
+		}
+	}
+
```
Comment (kurtisvg) [inline internal/lifecycle/readme_test.go]: nit: remove newline

File: internal/lifecycle/readme_test.go [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+	"strings"
+	"testing"
+)
+
+func setEnv(e map[string]string) error {
+	for k, v := range e {
+		err := os.Setenv(k, v)
+		if err != nil {
+			return err
+		}
```
Comment (kurtisvg) [inline internal/lifecycle/readme_test.go]: nit:
```suggestion
		if err := os.Setenv(k, v); err != nil {
			return err
		}
```

File: internal/lifecycle/readme_test.go [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+	"bufio"
+	"errors"
+	"os"
+	"os/exec"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+func setEnv(e map[string]string) error {
```
Comment (kurtisvg) [inline internal/lifecycle/readme_test.go:14]: nit: Comments are good, even on utility functions like this.

File: internal/lifecycle/readme_test.go [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+}
+
+func equalError(a, b error) bool {
+	if a == nil {
+		return b == nil
+	}
+	if b == nil {
+		return a == nil
+	}
+	return errors.Is(a, b)
```
Comment (kurtisvg) [inline internal/lifecycle/readme_test.go]: hmm, I'm not sure what this is doing, but it seems iffy. What are you trying to do?

File: internal/lifecycle/readme_test.go [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+	extractLifecycleErr  error
+	extractCodeBlocksErr error
+	env                  map[string]string
+	serviceName          string
+	gcrURL               string
+}
+
+var tests = []test{
+	{
+		in: "[//]: # ({sst-run-unix})\n" +
```
Comment (kurtisvg) [inline internal/lifecycle/readme_test.go]: Consider multiline strings or a single line - this feels a bit hard to read.

File: internal/lifecycle/readme_test.go [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+	}
+}
+
+func TestExtractCodeBlocks(t *testing.T) {
+	for i, tc := range tests {
+		s := bufio.NewScanner(strings.NewReader(tc.in))
+
+		h, err := extractCodeBlocks(s)
+		if !equalError(err, tc.extractCodeBlocksErr) {
+			t.Errorf("#%d: error mismatch\nhave: %v\nwant: %v", i, err, tc.extractCodeBlocksErr)
```
Comment (kurtisvg) [inline internal/lifecycle/readme_test.go]: Go has a common idiom for this: `want x, got y`

File: internal/lifecycle/readme_test.go [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+	toCommandsErr        error
+	extractLifecycleErr  error
+	extractCodeBlocksErr error
+	env                  map[string]string
+	serviceName          string
+	gcrURL               string
+}
+
+var tests = []test{
+	{
```
Comment (kurtisvg) [inline internal/lifecycle/readme_test.go:49]: I would consider adding a comment to each of these - what are these parameters supposed to be testing?

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: internal/lifecycle/readme_test.go (hunk 9 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go:49]: Resolved, thank you.

File: internal/lifecycle/readme_test.go (hunk 5 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go:14]: Resolved, thank you.

File: internal/lifecycle/readme_test.go (hunk 4 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go]: Resolved, thank you.

File: internal/lifecycle/readme_test.go (hunk 3 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go]: Resolved, thank you.

File: internal/lifecycle/readme.go (hunk 2 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme.go:125]: Resolved, thank you.

File: internal/lifecycle/readme_test.go (hunk 6 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go]: Resolved, thank you.

File: internal/lifecycle/readme_test.go (hunk 8 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go]: Resolved offline, thank you.

File: internal/lifecycle/readme.go (hunk 1 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme.go]: Resolved offline, thank you.

File: internal/lifecycle/readme_test.go (hunk 7 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go]: Resolved offline, thank you.

File: internal/lifecycle/readme_test.go (hunk 8 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go]: Resolved offline, thank you.

File: internal/lifecycle/readme_test.go (hunk 7 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go]: For reference: we're keeping this because it doesn't seem line multiline strings support escaped backticks.

**Review by kurtisvg — APPROVED** [review]

File: internal/lifecycle/readme_test.go [hunk 10]
```diff
@@ -0,0 +1,504 @@
...
+		cmds: []*exec.Cmd{
+			exec.Command("echo", "hello", "world"),
+			exec.Command("echo", "line", "one"),
+			exec.Command("echo", "line", "two"),
+		},
+	},
+}
+
+func TestToCommands(t *testing.T) {
+	for i, tc := range tests {
```
Comment (kurtisvg) [inline internal/lifecycle/readme_test.go]: Rather than iterate through each list of tests for each section, you should probably just have a different slice for each test.

**Review by kurtisvg — APPROVED** [review]

File: internal/lifecycle/readme_test.go (hunk 10 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go]: Resolved, thank you.

File: internal/lifecycle/readme_test.go [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+			t.Errorf("#%d: result mismatch\nwant: %#+v\ngot: %#+v", i, tc.cmds, cmds)
+		}
+
+		if err := unsetEnv(tc.env); err != nil {
+			t.Errorf("#%d: unsetEnv: %v", i, err)
+		}
+	}
+}
+
+type parseREADMETest struct {
```
Comment (kurtisvg) [inline internal/lifecycle/readme_test.go:233]: nit: Use MixedCaps

File: internal/lifecycle/readme_test.go [hunk 12]
```diff
@@ -0,0 +1,448 @@
...
+
+		// Cloud Run Service name and Container Registry URL tag replacement will be tested in TestToCommands
+		lifecycle, err := extractLifecycle(s, "", "")
+
+		if !errors.Is(err, tc.err) {
+			t.Errorf("#%d: error mismatch\nwant: %v\ngot: %v", i, tc.err, err)
+			continue
+		}
+
+		if err == nil && !reflect.DeepEqual(lifecycle, tc.lifecycle) {
```
Comment (kurtisvg) [inline internal/lifecycle/readme_test.go:324]: What if err != nil?

**Review by kurtisvg — APPROVED** [review]
2 nits, then LGTM

File: internal/lifecycle/readme_test.go (hunk 12 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go:324]: I wasn't sure if we wanted to still check the results if there was an error.

File: internal/lifecycle/readme_test.go (hunk 11 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go:233]: Resolved, thank you.

File: internal/lifecycle/readme_test.go (hunk 12 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go:324]: After talking offline, we're leaving this as is. Thank you.

**Review by dmahugh — APPROVED** [review]
LGTM

File: internal/lifecycle/lifecycle.go [hunk 13]
```diff
@@ -58,22 +58,20 @@ func NewLifecycle(sampleDir, serviceName, gcrURL string) (Lifecycle, error) {
 		readmePath = filepath.Join(sampleDir, "README.md")
 	}
 
-
-
 	if _, err := os.Stat(readmePath); err == nil {
-		lifecycle, err := parseREADME(readmePath, serviceName, gcrURL)
```
Comment (tritone) [inline internal/lifecycle/lifecycle.go:64]: All caps on README is actually appropriate here I think.

File: internal/lifecycle/readme.go [hunk 14]
```diff
@@ -42,7 +42,11 @@ var (
 
 	mdCodeFenceStartRegexp = regexp.MustCompile("^\\w*`{3,}[^`]*$")
 
-	errNoREADMECodeBlocksFound = fmt.Errorf("lifecycle.extractCodeBlocks: no code blocks immediately preceded by %s found", codeTag)
+	errNoReadmeCodeBlocksFound   = fmt.Errorf("lifecycle.extractCodeBlocks: no code blocks immediately preceded by %s found", codeTag)
```
Comment (tritone) [inline internal/lifecycle/readme.go:45]: I actually don't think this change is necessary, unless any of these errors are going to be used in more than one place. It makes it a little more complicated to read the code.

If anything, it might be easier for callers if there was a single error type that wrapped any error coming from readme parsing.

File: internal/lifecycle/readme.go [hunk 15]
```diff
@@ -68,7 +72,7 @@ func (cb codeBlock) toCommands(serviceName, gcrURL string) ([]*exec.Cmd, error)
 
 			i++
 			if i >= len(cb) {
-				return nil, fmt.Errorf("unexpected end of code block: expecting command line continuation; code block dump:\n%s", strings.Join(cb, "\n"))
+				return nil, fmt.Errorf("%w; code block dump:\n%s", errCodeBlockEndAfterLineCont, strings.Join(cb, "\n"))
```
Comment (tritone) [inline internal/lifecycle/readme.go]: The use of wrapping here doesn't really make sense-- usually that would just be when you are bubbling up an error from a called func.

File: internal/lifecycle/readme.go [hunk 16]
```diff
@@ -204,15 +218,15 @@ func replaceServiceName(command, serviceName string) string {
...
+	for i := 0; i < len(sp); i++ {
 		if sp[i] == os.ExpandEnv("$CLOUD_RUN_SERVICE_NAME") {
 			sp[i] = serviceName
 			return strings.Join(sp, " ")
 		}
 	}
 
 	// Searches for specific gcloud keywords and takes service name from them
-	for i := 0; i < len(sp) - 1; i++ {
+	for i := 0; i < len(sp)-1; i++ {
```
Comment (tritone) [inline internal/lifecycle/readme.go:229]: Were these formatting changes yours or from gofmt?

File: internal/lifecycle/readme_test.go [hunk 17]
```diff
@@ -0,0 +1,452 @@
...
+			exec.Command("gcloud", "--quiet", "run", "services", "update", uniqueServiceName, "--add-cloudsql-instances=project:region:instance"),
+		},
+		env: map[string]string{
+			"TEST_CLOUD_SQL_CONNECTION": "project:region:instance",
+		},
+	},
+}
+
+func TestToCommands(t *testing.T) {
+	for i, tc := range toCommandsTests {
```
Comment (tritone) [inline internal/lifecycle/readme_test.go:195]: Having a list of test cases and then looping through them seems like the right approach. I would also consider using a subtest for each case, which produces nicer output. This seems like it could work for most of your tests.

https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks
https://blog.golang.org/subtests

File: internal/lifecycle/readme_test.go [hunk 18]
```diff
@@ -0,0 +1,452 @@
...
+		},
+	},
+
+	// line cont char but code block closes at next line
+	{
+		codeBlock: codeBlock{
+			"echo multi \\",
+		},
+		cmds: nil,
+		err:  errCodeBlockEndAfterLineCont,
```
Comment (tritone) [inline internal/lifecycle/readme_test.go:87]: In general it's better not to have your test use unexported stuff from the package. Think about what kind of error a caller outside the package would expect to see. String matching is probably fine, or you could think about making an exported error that callers could actually check for.

Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go:87]: Resolved offline, thank you.

File: internal/lifecycle/readme.go (hunk 14 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme.go:45]: Resolved offline, thank you.

File: internal/lifecycle/readme.go (hunk 16 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme.go:229]: `gofmt`.

File: internal/lifecycle/readme_test.go (hunk 17 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go:195]: Creating an issue for this.

File: internal/lifecycle/lifecycle.go (hunk 13 above)
Comment (SaketramDurbha) [reply internal/lifecycle/lifecycle.go:64]: Resolved, thank you.

File: internal/lifecycle/readme.go (hunk 15 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme.go]: Resolved, thank you.

File: internal/lifecycle/readme_test.go (hunk 17 above)
Comment (SaketramDurbha) [reply internal/lifecycle/readme_test.go:195]: Actually, a PR: https://github.com/GoogleCloudPlatform/serverless-sample-tester/pull/31
